frontend/src/metabase/common/components/DataPicker/components/DataPickerModal.tsx

```diff
@@ -117,99 +120,127 @@ export const DataPickerModal = ({
     [onChange, onClose, tryLogRecentItem],
   );
 
-  const handleCardChange = useCallback(
-    (item: QuestionPickerItem) => {
-      if (!isValidValueItem(item.model)) {
-        return;
-      }
-
-      onChange(getQuestionVirtualTableId(item.id));
-      tryLogRecentItem(item);
-      onClose();
-    },
-    [onChange, onClose, tryLogRecentItem],
-  );
-
-  const tabs: EntityTab<NotebookDataPickerValueItem["model"]>[] = [
-    hasModels && hasNestedQueriesEnabled
-      ? {
-          displayName: t`Models`,
-          model: "dataset" as const,
-          icon: "model",
-          element: (
-            <QuestionPicker
-              initialValue={isModelItem(value) ? value : undefined}
-              models={MODEL_PICKER_MODELS}
-              options={options}
-              shouldShowItem={modelsShouldShowItem}
-              onItemSelect={handleCardChange}
-            />
-          ),
-        }
-      : undefined,
-    hasMetrics && hasNestedQueriesEnabled
-      ? {
-          displayName: t`Metrics`,
-          model: "metric" as const,
-          icon: "metric",
-          element: (
-            <QuestionPicker
-              initialValue={isMetricItem(value) ? value : undefined}
-              models={METRIC_PICKER_MODELS}
-              options={options}
-              shouldShowItem={metricsShouldShowItem}
-              onItemSelect={handleCardChange}
-            />
-          ),
-        }
-      : undefined,
-    {
-      displayName: t`Tables`,
-      model: "table" as const,
-      icon: "table",
-      element: (
-        <TablePicker
-          databaseId={databaseId}
-          value={isTableItem(value) ? value : undefined}
-          onChange={handleChange}
-        />
-      ),
-    },
-    hasQuestions && hasNestedQueriesEnabled
-      ? {
-          displayName: t`Saved questions`,
-          model: "card" as const,
-          icon: "folder",
-          element: (
-            <QuestionPicker
-              initialValue={isQuestionItem(value) ? value : undefined}
-              models={QUESTION_PICKER_MODELS}
-              options={options}
-              shouldShowItem={questionsShouldShowItem}
-              onItemSelect={handleCardChange}
-            />
-          ),
-        }
-      : undefined,
-  ].filter(
-    (tab): tab is EntityTab<NotebookDataPickerValueItem["model"]> =>
-      tab != null && models.includes(tab.model),
-  );
+  const tabs = useMemo(() => {
+    const computedTabs: EntityPickerTab<
```

Comment:
It's way easier to deal with types here when we're conditionally adding items to an array instead of filtering an array like before.